File: 1647.minimum-deletions-to-make-character-frequencies-unique.rb
#
# @lc app=leetcode id=1647 lang=ruby
#
# [1647] Minimum Deletions to Make Character Frequencies Unique
#

# @lc code=start
# @param {String} s
# @return {Integer}
def min_deletions(s)
  histogram = {}
  s.each_char { |c| histogram[c] = (histogram[c] || 0) + 1 }

  freq_arr = []
  histogram.each_value { |v| freq_arr[v] = (freq_arr[v] || 0) + 1 }

  subtractions = 0

  i = freq_arr.length
  until i.zero?
    if freq_arr[i] && freq_arr[i] > 1
      freq_arr[i - 1] = (freq_arr[i - 1] || 0) + freq_arr[i] - 1
      subtractions += freq_arr[i] - 1
    end
    i -= 1
  end

  subtractions
end
# @lc code=end

File: 1647.minimum-deletions-to-make-character-frequencies-unique.ts
/*
 * @lc app=leetcode id=1647 lang=typescript
 *
 * [1647] Minimum Deletions to Make Character Frequencies Unique
 */

// @lc code=start
function minDeletions(s: string): number {
  let histogram = {};
  let subtractions = 0;

  for (let i = 0; i < s.length; i++) {
    histogram[s.charCodeAt(i) % 26] =
      (histogram[s.charCodeAt(i) % 26] || 0) + 1;
  }

  let arr = [];
  for (let key in histogram) {
    arr[histogram[key]] = (arr[histogram[key]] || 0) + 1;
  }

  for (let i = arr.length - 1; i > 0; i--) {
    if (arr[i] > 1) {
      subtractions += arr[i] - 1;
      arr[i - 1] = (arr[i - 1] ?? 0) + (arr[i] - 1);
    }
  }

  return subtractions;
}
// @lc code=end
